File: languages/java/tests/src/test/java/co/revaly/sdk/ScrubberTests.java
package co.revaly.sdk;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import co.revaly.sdk.logging.RapScrubber;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

/** The single central allowlist scrubber (ADR-SDK-020): emit-known-safe, mask the rest. */
class ScrubberTests {

    @Test
    void allowlistedIdentifiersPassThroughVerbatim() {
        String scrubbed =
                RapScrubber.scrubJson(
                        "{\"merchantTransactionId\":\"mtx-1\",\"transactionStatus\":1,\"code\":\"not_processed\"}");

        assertTrue(scrubbed.contains("\"merchantTransactionId\":\"mtx-1\""));
        assertTrue(scrubbed.contains("\"transactionStatus\":1"));
        assertTrue(scrubbed.contains("\"code\":\"not_processed\""));
    }

    @Test
    void everythingElseIsScrubbedIncludingUnknownFutureFields() {
        String scrubbed =
                RapScrubber.scrubJson(
                        "{\"cardNumber\":\"4111111111111111\",\"cvv\":\"123\",\"someNewField\":\"value\"}");

        assertFalse(scrubbed.contains("4111111111111111"));
        assertFalse(scrubbed.contains("123"));
        assertFalse(scrubbed.contains("value"));
        assertTrue(scrubbed.contains(RapScrubber.SCRUBBED));
    }

    @Test
    void nestedStructureIsPreservedAndScrubbedRecursively() {
        String scrubbed =
                RapScrubber.scrubJson(
                        "{\"paymentMethod\":{\"creditCard\":{\"cardNumber\":\"4111111111111111\"},"
                                + "\"transactionId\":\"txn-1\"}}");

        assertFalse(scrubbed.contains("4111111111111111"));
        assertTrue(scrubbed.contains("\"transactionId\":\"txn-1\""));
        assertTrue(scrubbed.contains("\"paymentMethod\""));
        assertTrue(scrubbed.contains("\"creditCard\""));
    }

    @Test
    void arrayScalarsFollowTheParentKeysStatus() {
        String scrubbed =
                RapScrubber.scrubJson("{\"attempts\":[1,2],\"phoneNumbers\":[\"555-0100\"]}");

        assertTrue(scrubbed.contains("\"attempts\":[1,2]"));
        assertFalse(scrubbed.contains("555-0100"));
    }

    @Test
    void unparseableInputNeverEchoesTheRawText() {
        assertEquals("[unparseable:scrubbed]", RapScrubber.scrubJson("PAN=4111111111111111"));
        assertEquals("", RapScrubber.scrubJson(null));
        assertEquals("", RapScrubber.scrubJson("  "));
    }

    @Test
    void authorizationHeaderIsAlwaysRedacted() {
        Map<String, List<String>> headers = new HashMap<>();
        headers.put("Authorization", Collections.singletonList("ApiKey secret-key"));
        headers.put("Content-Type", Collections.singletonList("application/json"));
        headers.put("X-Correlation-ID", Collections.singletonList("corr-1"));
        headers.put("X-Custom-Internal", Collections.singletonList("internal-value"));

        Map<String, String> scrubbed = RapScrubber.scrubHeaders(headers);

        assertEquals(RapScrubber.REDACTED, scrubbed.get("Authorization"));
        assertEquals("application/json", scrubbed.get("Content-Type"));
        assertEquals("corr-1", scrubbed.get("X-Correlation-ID"));
        assertEquals(RapScrubber.REDACTED, scrubbed.get("X-Custom-Internal"));
        assertFalse(scrubbed.toString().contains("secret-key"));
    }

    @Test
    void headerLookupIsCaseInsensitive() {
        Map<String, List<String>> headers = new HashMap<>();
        headers.put("authorization", Collections.singletonList("ApiKey secret-key"));

        Map<String, String> scrubbed = RapScrubber.scrubHeaders(headers);

        assertEquals(RapScrubber.REDACTED, scrubbed.get("Authorization"));
    }
}

File: languages/dotnet/tests/Revaly.Sdk.Tests/ScrubberTests.cs
using Revaly.Sdk.Logging;
using Revaly.Sdk.Testing;

namespace Revaly.Sdk.Tests;

/// <summary>
/// The central allowlist scrubber (ADR-SDK-020): tested against a full charge payload,
/// not a field list that can drift — unknown fields must fail safe (scrubbed).
/// </summary>
public class ScrubberTests
{
    [Fact]
    public void Card_material_never_survives_scrubbing()
    {
        var payload = $"{{\"amount\":1999,\"merchantTransactionId\":\"{SyntheticData.MerchantTransactionId}\"," +
            $"\"currency\":\"USD\",\"paymentMethod\":{{\"creditCard\":{{\"number\":\"{SyntheticData.TestPan}\"," +
            $"\"expiryMonth\":\"12\",\"expiryYear\":\"2030\",\"cardVerificationCode\":\"{SyntheticData.TestCvv}\"}}}}," +
            "\"customerIp\":\"203.0.113.9\",\"description\":\"synthetic order\"}";

        var scrubbed = RapScrubber.ScrubJson(payload);

        Assert.DoesNotContain(SyntheticData.TestPan, scrubbed, StringComparison.Ordinal);
        Assert.DoesNotContain(SyntheticData.TestCvv, scrubbed, StringComparison.Ordinal);
        Assert.DoesNotContain("203.0.113.9", scrubbed, StringComparison.Ordinal);
        Assert.DoesNotContain("1999", scrubbed, StringComparison.Ordinal);
        Assert.Contains(SyntheticData.MerchantTransactionId, scrubbed, StringComparison.Ordinal);
        Assert.Contains("USD", scrubbed, StringComparison.Ordinal);
        Assert.Contains(RapScrubber.Scrubbed, scrubbed, StringComparison.Ordinal);
    }

    [Fact]
    public void Unknown_future_fields_fail_safe()
    {
        var scrubbed = RapScrubber.ScrubJson(
            "{\"someNewField\":\"secret-value\",\"transactionId\":\"txn_1\"}");

        Assert.DoesNotContain("secret-value", scrubbed, StringComparison.Ordinal);
        Assert.Contains("txn_1", scrubbed, StringComparison.Ordinal);
    }

    [Fact]
    public void Non_json_input_returns_a_placeholder_never_the_raw_text()
    {
        var scrubbed = RapScrubber.ScrubJson("PAN=4111111111111111 raw text");

        Assert.DoesNotContain("4111111111111111", scrubbed, StringComparison.Ordinal);
        Assert.Equal("[unparseable:scrubbed]", scrubbed);
    }

    [Fact]
    public void Authorization_headers_are_redacted_even_in_traces()
    {
        var headers = new Dictionary<string, IEnumerable<string>>
        {
            ["Authorization"] = new[] { $"ApiKey {SyntheticData.TestApiKey}" },
            ["X-Api-Version"] = new[] { "2.1" },
            ["X-Custom"] = new[] { "anything" },
        };

        var scrubbed = RapScrubber.ScrubHeaders(headers);

        Assert.Equal(RapScrubber.Redacted, scrubbed["Authorization"]);
        Assert.Equal("2.1", scrubbed["X-Api-Version"]);
        Assert.Equal(RapScrubber.Redacted, scrubbed["X-Custom"]);
    }

    [Fact]
    public void Arrays_and_nested_structures_are_preserved_but_scrubbed()
    {
        var scrubbed = RapScrubber.ScrubJson(
            "{\"items\":[{\"number\":\"4111111111111111\"},{\"transactionId\":\"txn_2\"}]}");

        Assert.DoesNotContain("4111111111111111", scrubbed, StringComparison.Ordinal);
        Assert.Contains("txn_2", scrubbed, StringComparison.Ordinal);
    }
}
